opentelemetry/proto/collector/logs/v1/logs_service.proto

```diff
@@ -43,4 +43,35 @@ message ExportLogsServiceRequest {
 }
 
 message ExportLogsServiceResponse {
+  // The details of a partially successful export request.
+  // If the processing of the request is partially successful
+  // (i.e. when the server accepts only parts of the data and rejects the rest)
+  // and the number of accepted log records is known, the server MUST initialize
+  // the partial_success field and MUST set the accepted_log_records.
+
+  // When the export request is processed successfully, the server MUST leave
+  // the partial_success field unset.
+  ExportLogsPartialSuccess partial_success = 1;
+}
+
+message ExportLogsPartialSuccess {
+  // The number of accepted log records
+  //
+  // If the request is partially successful and the number of
+  // accepted log records is known, the server MUST set
```

Comment:
What is the situation when the number of accepted log records is unknown and why is it useful for the sender to know about this situation specifically? If there is no such situation I suggest that we get rid of this and require the server always sets this number (if it doesn't know then set it either to 0 or to total number - either way is fine). Doing so will eliminate the need to have optional accepted_log_records (optional fields are poorly supported in some Protobuf libraries and are a pain to deal with).

Reply:
>What is the situation when the number of accepted log records is unknown and why is it useful for the sender to know about this situation specifically

I also tend to think this shouldn't happen. If the server doesn't know the number of accepted "things", it should return an empty response, without populating the `partial_success` field, just as it is done today.

The discussion around `optional` was without it, a server could initialize the `partial_success` property, set the `error_message` and leave the `accepted_<signal>` field with the default value, which is `0`. In this situation, it would not be possible to differentiate a `0` being the the default/unset from the server actually dropping everything.

@jmacd had some other alternatives to this here: https://github.com/open-telemetry/opentelemetry-proto/pull/390#discussion_r916322737

Reply:
>require that the accepted_<signal> field value is always set (if partial_success is set)

Yes, I agree and that's what I wrote in the spec now. Although we still have the issue where a sender receives a `partial_success` with something in the error message and `0` for `accepted_<signal>` and it then must interpret it as "nothing was accepted" which is clear to me, but could be open to other interpretations, since `0` is also the not set/default value.

Reply:
I still see `optional` qualifier for `accepted_log_records` field. Do you plan to remove it or you disagree with my opinion that we don't need it?

Reply:
I didn't remove the `optional` in any for now. I wanted to see what others think before changing it, specially since there is this "nuance" with the default/not set value being the same thing. I know from talking with @jmacd that he had concerns with this double meaning. I think I agree with you that if the spec says to always populate it, if you get a `0` it means nothing was accepted so I'm in favor of removing it.

Reply:
> IMO, we should: get rid of the optional for accepted_<signal> field require that the accepted_<signal> field value is always set (if partial_success is set). There is no such thing as ...

The problem is that "partial success is set" cannot be answered because without `optional` there is no such thing as field presence for messages. If the server responds with a partial_success message having 0 for the accepted_things field and an empty message, the receiver will see an unset partial_success. It should be possible to meaningfully indicate that 0 points were accepted, and we can't rely on field presence without `optional`. Otherwise, we need to infer that partial_success was "set" because it has a non-default field. 

My goal behind supporting this PR is to get to a place where clients have exact information about how many points are being successfully sent vs. dropped. There should be no ambiguity. Moreover, we should not have to fill in a struct when 100% of points are accepted, because it creates an extra burden on the sender. In order to have a meaningful zero and gain the performance benefit without `optional`, we need to change again this PR to count "dropped" things instead of "accepted" things.

I support removing `optional`, but all the "accepted_" fields should revert back to "dropped_" fields. The 0 value will always be meaningful, then.

Reply:
I can do the change and submit a new PR if that's what everyone agrees, no problem. 

But I wonder, if we will not have the same dilemma again. If the server responds with a `partial_success` with `rejected_<signal>` = `0` and `error_message` = `"` what does that mean? Nothing was dropped? Then why did the server even initialized the `partial_success`, which it shouldn't?  Then we can argue that we will write in the spec that this shouldn't happen but isn't it what is already written. just the other way around?

> It should be possible to meaningfully indicate that 0 points were accepted

The current spec change says if everything was accepted, `partial_success` should **not** be set, so I feel it's already enough if you get a `partial_success` and `accepted_<signal>` is `0` then nothing was accepted. Sure, with `optional` there's _more_ certainty but without it, I feel we are also covered, like @tigrannajaryan mentioned. 

Moreover, we can't enforce what servers will do, only recommend what it _should_ do. I feel with what we have written in the spec, we already have it covered. If it's important to me that clients get this right,

Reply:
> If the server responds with a partial_success with rejected_<signal> = 0 and error_message = " what does that mean? Nothing was dropped? Then why did the server even initialized the partial_success, which it shouldn't?

There is effectively no test that distinguishes "server initialized the partial_success" vs not when the default-value is used, short of using `optional`.

Without `optional`, there are two choices:
- We either can infer that the number-value is meaningful because the string-value is not empty
- We can declare that the number-value is always meaningful. 

We cannot differentiate a field that is set to the default value from a field that was not set. To your question, when the partial success field is not set, as it will not be for older OTLP versions, I will assume nothing dropped. I do not believe there is any other valid interpretation, since we have not yet specified any form of partial success. Until we specify partial success, you have all-or-none semantics.

Reply:
To me this is reasonable semantics:
1. If `partial_success` is not set then it means the request is processed fully successfully. This also happens to be the old behavior, so no changes in the semantics, which is good - we are backward compatible.
2. If `partial_success` is set then it means something went wrong. The `accepted_<signal>` field inside `Export*PartialSuccess` message MUST be set. 0 is a valid value for `accepted_<signal>` field and means all records failed to process. `error_message` MUST be also set to indicate what the problem was. Empty string happens to be a valid value, meaning that there is no _additional_ error text to provide, however the entire situation is still erroneous.

I think this is all we need, it does not need to be more complicated than this.

How this is used in the Server:
1. If the Server processed everything successfully then don't set `partial_success`.
2. If processing of any record failed the set `partial_success` and calculate `accepted_<signal>` field value as `total records in the record`-`failed to process record`. Again, 0 is perfectly fine value to end up with. Also set `error_message` to indicate the failure reason (can be as simple as "some metric data points are invalid" or more complex if desirable, e.g. one line per failed record with precise failure reason).

How this is used in the Client:
1. If `partial_success` in the response is not set, good, nothing else to do.
2. If `partial_success` in the response is set then print an error message in the log and include `error_message`. Also log `accepted_<signal>` value. Can also keep a metric based on `accepted_<signal>` value if desirable.

I may be missing something here. Perhaps there is a scenario that is not covered here and where the Server or the Client need to make a different decision and behave differently?